#' ICAMS: In-depth Characterization and Analysis of Mutational Signatures
#'
#' A toolkit for analysis and visualization
#' of experimentally elucidated mutational signatures -- the kind of analysis
#' and visualization presented in Boot et al., "In-depth characterization of
#' the cisplatin mutational signature in human cell lines and in esophageal and
#' liver tumors", \emph{Genome Research}, 2018,
#' https://genome.cshlp.org/content/28/5/654.short.
#'
#' ICAMS can read in variant call files (VCFs)
#' generated by Strelka or Mutect, and collate the
#' mutations  into "catalogs" of mutational
#' spectra. ICAMS can plot the catalogs of mutational spectra
#' and signatures.
#'
#' ICAMS can build and plot catalogs of mutational spectra
#' for single nucleotide
#' substitutions (SNS), double nucleotide substitutions (DNS),
#' and small insertions and deletions (ID). It can also read and
#' write these catalogs.
#'
#' @section Creating catalogs from from variant call files (VCF files):
#' \enumerate{
#' \item \code{\link{StrelkaSNSVCFFilesToCatalog}}, which creates 3 SNS catalogs (96,
#' 192, 1536) and 3 DNS catalogs (78, 136, 144) from the Strelka SNS VCFs.
#'
#' \item \code{\link{StrelkaIDVCFFilesToCatalog}}, which creates ID (indels) catalog
#' from the Strelka ID VCFs.
#'
#' \item \code{\link{MutectVCFFilesToCatalog}}, which creates 3 SNS catalogs (96,
#' 192, 1536), 3 DNS catalogs (78, 136, 144) and ID (indels) catalog from the
#' Mutect VCFs.
#' }
#'
#' @section Plotting catalogs:
#' Functions for plotting catalogs of mutational spectra or
#' of mutational signatures to a PDF file. Mutational \emph{signatures}
#' are similar to spectra, but where spectra consist of
#' counts of mutations in each mutation class
#' (e.g. ACA > AAA, ACA > AGA, ACA > ATA, ACC > AAC, ...)
#' signatures consist of
#' the proportions of mutations in each class (with all the
#' proportions summing to 1).
#' \code{\link{PlotCatalogToPdf}}
#'
#' @section Writing catalogs:
#' Functions for writing a catalog of mutational spectra or of
#' mutational signatures to a file on disk.
#' \code{\link{WriteCatalog}}
#'
#' @section Reading catalogs:
#' Functions for reading files that contain catalogs of mutational
#' spectra or of signatures in standardized format.
#' \code{\link{ReadCatalog}}
#'
#' @section Transforming catalogs:
#' There is a function to transform catalogs of mutational spectra or
#' signatures to account for differing abundances of the source
#' sequence of the mutations in the genome.  For example, mutations from
#' ACG are much rarer in the human genome than mutations from ACC
#' simply because CG dinucleotides are rare in the genome.
#' This function can also transform spectra
#' based on observed genome-wide counts to "density"-based
#' catalogs. In density-based catalogs
#' mutations are expressed as mutations per
#' (million) source sequences. For example,
#' a density-based catalog represents
#' the proportion of ACCs mutated to
#' ATCs, the proportion of ACGs mutated to ATGs, etc.
#' This is
#' opposed count-based catalogs, which
#' contain the number of ACC-to-ATC mutations, the number of
#' ACG-to-ATG mutations, etc.
#' This function can also transform observed-count based
#' spectra or signatures from genome to exome based counts,
#' or between different species (since the abundances of
#' source sequences vary between genome and exome and between
#' species).
#' \code{\link{TransformCatalog}}
#'
#' @section Collapsing catalogs:
#' Functions for collapsing a mutational spectrum or signature catalog
#' based on a fined-grained set of features (for example, single-nucleotide
#' substitutions in the context of the preceding and following 2 bases)
#' to a catalog based on a coarser-grained set of features
#' (for example, single-nucleotide substitutions in the context of the
#' immediately precedig and following bases).
#' \code{\link{CollapseCatalog}}
#'
#' @section Exported data: \enumerate{
#'
#' \item \code{\link{CatalogRowOrder}} Canonical order of row names in a catalog.
#'
#'\item \code{\link{TranscriptRanges}} Transcript ranges and strand information
#'for a particular organism. }
#' @docType package
#' @name ICAMS
NULL
